### PR TITLE
State select

### PR DESCRIPTION
**Pull request by tandronescu**
Final working code for state select, with code standards, comments and working launch file with turtle sim available to actually be controlled by the xbox controller. Change the subscriber in the drive_lis_pub file to the actual rover when necessary.

### CODE REVIEW

File: Workspace/src/xbox_controller/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+# if(TARGET ${PROJECT_NAME}-test)
+#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
+# endif()
+
+## Add folders to be run by python nosetests
+# catkin_add_nosetests(test)
+
+catkin_package()
+
+add_executable(xbox teleop_node.cpp)
```
Comment (lijerryli) [inline Workspace/src/xbox_controller/CMakeLists.txt]: can you change the xbox node name to state_select to more accurately describe the node

File: Workspace/src/xbox_controller/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+add_executable(state drive_lis_sub.cpp)
```
Comment (lijerryli) [inline Workspace/src/xbox_controller/CMakeLists.txt]: and change this state node to be named "joy_to_twist" for accuracy

File: Workspace/src/xbox_controller/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+## Compile as C++11, supported in ROS Kinetic and newer
+# add_compile_options(-std=c++11)
+
+## Find catkin macros and libraries
+## if COMPONENTS list like find_package(catkin REQUIRED COMPONENTS xyz)
+## is used, also find other catkin packages
+find_package(catkin REQUIRED COMPONENTS
+  roscpp
+  rospy
+  std_msgs
```
Comment (lijerryli) [inline Workspace/src/xbox_controller/CMakeLists.txt:13]: add sensor_msgs as a required component (should be standard on all ROS installs but just in case)

Comment (lijerryli) [reply Workspace/src/xbox_controller/CMakeLists.txt:13]: same with geometry_msgs (for twist)

File: Workspace/src/xbox_controller/launch/state_select.launch [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<launch>
+	<node name="teleop_node" pkg="xbox_controller" type="xbox" />
+	<node name="drive_lis_pub" pkg="xbox_controller" type="state" />
+	<node name="joy_launcher" pkg="joy" type="joy_node" args="usb-©Microsoft_Corporation_Controller_15D5CD4-joystick" />
```
Comment (lijerryli) [inline Workspace/src/xbox_controller/launch/state_select.launch]: does the argument to select the xbox controller not require
`<param name="dev" type="string" value="/dev/input/by-id/usb-©Microsoft_Corporation_Controller_15D5CD4-joystick" />`
Look at http://wiki.ros.org/joy/Tutorials/WritingTeleopNode#Writing_a_Launch_File_to_Start_all_of_the_Nodes

File: Workspace/src/xbox_controller/teleop_node.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
+#include "ros/ros.h"
+#include "sensor_msgs/Joy.h"
+#include "std_msgs/String.h"
+#include "teleop_twist.h"
```
Comment (lijerryli) [inline Workspace/src/xbox_controller/teleop_node.cpp]: This node should not be including teleop_twist.h

**Review by lijerryli — DISMISSED** [review]

File: Workspace/src/xbox_controller/launch/state_select.launch [hunk 6]
```diff
@@ -0,0 +1,23 @@
+<launch>
+	<node name="teleop_node" pkg="xbox_controller" type="state_select" />
+	
+	<node name="drive_lis_pub" pkg="xbox_controller" type="joy_to_twist" />
+	
+	<node name="joy_launcher" pkg="joy" type="joy_node">
+	
+	<param name="controller" type="string" value="/dev/input/by-id/usb-©Microsoft_Corporation_Controller_15D5CD4-joystick" />
```
Comment (lijerryli) [inline Workspace/src/xbox_controller/launch/state_select.launch]: I think `name` has to be `"dev"` for this to work. Also just change the formatting a bit to make it a bit easier to read:
```
<node name="joy_launcher" pkg="joy" type="joy_node"> 
    <param name="dev" type="string" value="/dev/input/by-id/usb-©Microsoft_Corporation_Controller_15D5CD4-joystick" /> 
</node>

File: Workspace/src/xbox_controller/launch/state_select.launch [hunk 7]
```diff
@@ -0,0 +1,21 @@
+<launch>
+	<node name="teleop_node" pkg="xbox_controller" type="state_select" />
+	
+	<node name="drive_lis_pub" pkg="xbox_controller" type="joy_to_twist" />
+	
+	<node name="joy_launcher" pkg="joy" type="joy_node">
+		<param name="dev" type="string" value="/dev/input/by-id/usb-©Microsoft_Corporation_Controller_15D5CD4-joystick" />
```
Comment (archielee) [inline Workspace/src/xbox_controller/launch/state_select.launch]: @lijerryli this looks ugly af. should be a udev rule right?

File: Workspace/src/xbox_controller/drive_lis_sub.cpp [hunk 8]
```diff
@@ -0,0 +1,201 @@
+#include "geometry_msgs/Twist.h"
```
Comment (archielee) [inline Workspace/src/xbox_controller/drive_lis_sub.cpp]: can you put your code inside a folder called "src"

File: Workspace/src/xbox_controller/teleop_twist.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
+/**
```
Comment (archielee) [inline Workspace/src/xbox_controller/teleop_twist.h]: This should be inside a folder called "include". You may need to update the CMakeLists.txt file (specifically the line that says `include_directories`)

**Review by archielee — CHANGES_REQUESTED** [review]

File: Workspace/src/xbox_controller/launch/state_select.launch (hunk 7 above)
Comment (lijerryli) [reply Workspace/src/xbox_controller/launch/state_select.launch]: no, the xbox controller didn't work with udev rules due to the wireless dongle supporting 4 controllers. This was the only way to select it.

Comment (archielee) [reply Workspace/src/xbox_controller/launch/state_select.launch]: right this is the wireless controller. we should default to the wired controller (js0) and only use the wireless if required

Comment (tandronescu) [reply Workspace/src/xbox_controller/launch/state_select.launch]: so you want me to keep the id argument or switch it back to js0?

**Review by archielee — APPROVED** [review]

**Pull request merged**